### PR TITLE
[TOSA] Update type converter and unary ops

### PR DESCRIPTION
**Pull request by BraunPhilipp**
This PR serves the following purposes:
* Update type converter to legalize valid TOSA types
* Restructure `ONNXToTOSA` pass to align with other conversion passes
* Support for `onnx.Floor` and `onnx.Neg`
* LIT tests for unary ops

Signed-off-by: Philipp Braun <philipp.braun@amd.com>

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp
Comment (sjarus) [inline src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:67]: A subtle distinction between Floor and Negate is that Floor is defined in the TOSA spec only for fp types. Negate is defined for both integer and fp. 
https://www.mlplatform.org/tosa/tosa_spec.html#_floor
https://www.mlplatform.org/tosa/tosa_spec.html#_negate

This is currently noted as a bug in the TorchToTosa path as well - I remembered to check since I'd implemented that piece myself.

**Review by sjarus — CHANGES_REQUESTED** [review]
Thanks for this starting contribution @BraunPhilipp! 

The TOSA parts look great, except for the suggestion inline.

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp
Comment (KFAFSP) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:67]: We could either not use the macro mechanism and have a dedicated pattern, or provide an explicit template specialization here. In either case, it seems appropriate to use the same `notifyMatchFailure` approach on integer types to floor for now.

Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:67]: We use a few macros here and there, generally we try to avoid them and use templated functions or some other C++ mechanisms.

**Review by AlexandreEichenberger — COMMENTED** [review]
@chentong319 this pr use type conversions, since you just worked with this, you may want to give a look to this PR too... thanks

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp
Comment (BraunPhilipp) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:86]: Thank you for the quick response! By my understanding `onnx.Floor` will never use an integer type in `onnx-mlir` https://github.com/onnx/onnx/blob/main/docs/Operators.md#Floor. Nevertheless, I have updated the type converter. Hope this is slightly clearer.

**Comment by AlexandreEichenberger** [comment]
@jenkins-droid test this please

**Comment by BraunPhilipp** [comment]
@sjarus @AlexandreEichenberger I tried to address the comments above in the last commit. Not quite sure if it is possible to disable droid messages; they seem a bit disruptive. The current commits are again rebased on top of main; hopefully the right procedure.

**Comment by sjarus** [comment]
> Thanks again @sjarus @AlexandreEichenberger for reviewing! I tried to address the comments (`notifyMatchFailure` + removal of macros) in the last commit. Not quite sure if it is possible to disable droid messages; they seem a bit disruptive. The current commits are again rebased on top of main; hopefully the right procedure. Please let me know in case this PR requires other changes.

The new version looks good functionally! But I think you may find in future that when similar fp-only ops are added - CeilOp for example - you'd be duplicating the specialization that would only differ in the src/dst op names.

**Comment by AlexandreEichenberger** [comment]
@BraunPhilipp please familiarize yourself with DCO. If you state here compliance with the policy, I will override the check for this PR, no need to do a new one.

From the bottom of our README:

ONNX-MLIR requires committers to sign their code using the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). Practically, each git commit needs to be signed, see [here](https://github.com/onnx/onnx-mlir/blob/main/docs/Workflow.md#step-7-commit--push) for specific instructions.

**Comment by AlexandreEichenberger** [comment]
PS: we have some network issues that result sometimes in jenkins failures. I will run it again, your PR is likely fine.

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp
Comment (AlexandreEichenberger) [inline src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:23]: This template works well and I am not asking you to change the code. 

You could still look at the template pattern we use in src/Conversion/Math/Elementwise.cpp where we use this:

```C++
template <>
struct ScalarOp<ONNXDivOp> {
  using FOp = arith::DivFOp;
  using IOp = arith::DivSIOp;
};
```
to reduce a bit on the redundant code that you have here. Let me know if you want to change it to that pattern. If not, I am fine accepting the PR as is with the understanding that if you add many more element wise op you would at least consider this pattern, if it fits your usage pattern.

**Review by AlexandreEichenberger — APPROVED** [review]
LGTM, again let me know if you want to change the template pattern or not.

**Comment by AlexandreEichenberger** [comment]
@BraunPhilipp please let me know if you want to further change the code, or if you want the code to land as is.

In all cases, you  need to confirm here that you abided by the DCO requirements. Once you do so, I can override the DCO failure when merging with main.

File: src/Conversion/ONNXToTOSA/Math/Elementwise.cpp
Comment (BraunPhilipp) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:23]: > @BraunPhilipp please let me know if you want to further change the code, or if you want the code to land as is.
> 
> In all cases, you need to confirm here that you abided by the DCO requirements. Once you do so, I can override the DCO failure when merging with main.

Sorry for the late reply due to Europe's vacation season. Thank you for the template suggestion! I will update it this week and double check the DCO requirements.

Comment (AlexandreEichenberger) [reply src/Conversion/ONNXToTOSA/Math/Elementwise.cpp:23]: From Swizerland myself... vacations are important :-)

**Comment by BraunPhilipp** [comment]
@Eichenberg Sorry for the late reply! This turned out to be legally more challenging than expected. 

> I hereby confirm that I abided by the DCO requirements.

For the PR itself, I just rebased on top of main and tweaked the templating. Let me know if the implementation requires further changes. Further contributions should be smoother.

**Comment by AlexandreEichenberger** [comment]
updated branch to latest and started the CIs

**Pull request merged**
